README.md
# ez-vcard

|     |     |
| --- | --- |
| Continuous Integration: | [![](https://travis-ci.org/mangstadt/ez-vcard.svg?branch=master)](https://travis-ci.org/mangstadt/ez-vcard) |
| Code Coverage: | [![codecov.io](http://codecov.io/github/mangstadt/ez-vcard/coverage.svg?branch=master)](http://codecov.io/github/mangstadt/ez-vcard?branch=master) |
| Maven Central: | [![Maven Central](https://maven-badges.herokuapp.com/maven-central/com.googlecode.ez-vcard/ez-vcard/badge.svg)](https://maven-badges.herokuapp.com/maven-central/com.googlecode.ez-vcard/ez-vcard) |
| Chat Room: | [![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/mangstadt/ez-vcard?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge) |
| License: | [![FreeBSD License](https://img.shields.io/badge/License-FreeBSD-red.svg)](https://github.com/mangstadt/ez-vcard/blob/master/LICENSE) |

ez-vcard is a vCard library written in Java.  It can read and write vCards in many different formats.  The "ez" stands for "easy" because the goal is to create a library that's easy to use.

<p align="center"><strong><a href="https://github.com/mangstadt/ez-vcard/wiki/Downloads">Downloads</a> |
<a href="http://mangstadt.github.io/ez-vcard/javadocs/latest/index.html">Javadocs</a> |
<a href="#mavengradle">Maven/Gradle</a> | <a href="https://github.com/mangstadt/ez-vcard/wiki">Documentation</a></strong></p>

```java
String str =
"BEGIN:VCARD\r\n" +
"VERSION:4.0\r\n" +
"N:Doe;Jonathan;;Mr;\r\n" +
"FN:John Doe\r\n" +
"END:VCARD\r\n";

VCard vcard = Ezvcard.parse(str).first();
String fullName = vcard.getFormattedName().getValue();
String lastName = vcard.getStructuredName().getFamily();
```

```java
VCard vcard = new VCard();

StructuredName n = new StructuredName();
n.setFamily("Doe");
n.setGiven("Jonathan");
n.getPrefixes().add("Mr");
vcard.setStructuredName(n);

vcard.setFormattedName("John Doe");

String str = Ezvcard.write(vcard).version(VCardVersion.V4_0).go();
```

# Features

 * Simple, intuitive API (see [Examples](https://github.com/mangstadt/ez-vcard/wiki/Examples)).
 * Android compatibility (Requires android api 21. DateTime (i.e. Birthday), TimeZone and java.nio.file.Path Requires either Android-API-26 or desuaring.) 
 * Full compliance with 2.1, 3.0, and 4.0 specifications (see [Supported Specifications](https://github.com/mangstadt/ez-vcard/wiki/Supported-Specifications)).
 * Supports XML, HTML, and JSON encoded vCards (see [Supported Specifications](https://github.com/mangstadt/ez-vcard/wiki/Supported-Specifications)).
 * Extensive unit test coverage.
 * Low Java version requirement (1.8 or above).
 * Few dependencies on external libraries.  Dependencies can be selectively excluded based on the functionality that is needed (see [Dependencies](https://github.com/mangstadt/ez-vcard/wiki/Dependencies)).
 * Can be build either with maven or gradle.
 * Can be included as a gradle-Subproject.

# News

**August 27, 2023**

[Version 0.12.1](https://github.com/mangstadt/ez-vcard/wiki/Downloads) released.  This release includes a few bug fixes.  Please see the [changelog](https://github.com/mangstadt/ez-vcard/wiki/Changelog) for details.

**February 4, 2023**

[Version 0.12.0](https://github.com/mangstadt/ez-vcard/wiki/Downloads) released.  This release updates to Java 1.8 and a few other things.  Please see the [changelog](https://github.com/mangstadt/ez-vcard/wiki/Changelog) for details.

**September 11, 2021**

[Version 0.11.3](https://github.com/mangstadt/ez-vcard/wiki/Downloads) released.  This release fixes several issues, including a security vulnerability.  Please see the [changelog](https://github.com/mangstadt/ez-vcard/wiki/Changelog) for details.

[Old News](https://github.com/mangstadt/ez-vcard/wiki/Old-News)

# Maven/Gradle

**Maven**

```xml
<dependency>
   <groupId>com.googlecode.ez-vcard</groupId>
   <artifactId>ez-vcard</artifactId>
   <version>0.12.1</version>
</dependency>
```

**Gradle**

```
compile 'com.googlecode.ez-vcard:ez-vcard:0.12.1'
```

# Build Instructions via maven pom.xml

ez-vcard uses [Maven](http://maven.apache.org/) as its build tool, and adheres to its conventions.

To build the project: `mvn compile`  
To run the unit tests: `mvn test`  
To build a JAR: `mvn package`

# Build Instructions via gradle build.gradle
To build the lib project: `./gradlew jar`  
To run the unit tests: `./gradlew test`

**Eclipse users:** Due to a quirk in the build process, before running the `eclipse:eclipse` goal, you must tweak some of the `<resource>` definitions in the POM file.  See the comments in the POM file for details.

# Questions / Feedback

You have some options:

 * [Issue tracker](https://github.com/mangstadt/ez-vcard/issues)
 * [Gitter chat room](https://gitter.im/mangstadt/ez-vcard)
 * [Post a question to StackOverflow](http://stackoverflow.com/questions/ask) with `vcard` as a tag.
 * Email me directly: [mike.angstadt@gmail.com](mailto:mike.angstadt@gmail.com)

Please submit bug reports and feature requests to the [issue tracker](https://github.com/mangstadt/ez-vcard/issues).  Contributors are listed in the project credits.

# Java 6 branch

For those who can't use Java 8 (i.e. older Android APIs), there exists a parallel, Java 6 branch. This branch receives the same bug fixes that the main branch receives.

https://github.com/mangstadt/ez-vcard/tree/j6

# Android

The [ez-vcard-android project](https://github.com/mangstadt/ez-vcard-android) interfaces between ez-vcard and the native Android APIs.  Note, however, that it is not being actively maintained.

# Node.js

A Node.js wrapper for ez-vcard can be found here: https://github.com/Maxim-Mazurok/ez-vcard

# Credits

**Lead Developer**  
Michael Angstadt

**Documentation**  
Michael Angstadt

**Architecture Ideas**  
George El-Haddad ([CardMe Project](https://sourceforge.net/projects/cardme/))

**Maven Central Reviewer**  
Joel Orlina

**Project Hosting**  
[Github](https://github.com)  
[Google Code](https://code.google.com)

**Contributors**  
amarnathr ([hCard template bug](https://github.com/mangstadt/ez-vcard/issues/16))  
[bgorven](https://github.com/bgorven) (jackson-databind integration, jackson pretty printing, [round-trip testing](https://github.com/mangstadt/ez-vcard/pull/49))  
Moritz Bechler (Geo URI bug fix)  
Kiran Kumar Bhushan (quoted-printable bug)  
[Sean Boylan](https://github.com/seanboylan) ([XXE vulnerability](https://github.com/mangstadt/ez-vcard/issues/55))  
Florian Brunner ([OSGi metadata](https://github.com/mangstadt/ez-vcard/issues/11))  
Pratyush Chandra ([ez-vcard-android](http://github.com/mangstadt/ez-vcard-android))  
Lívio Cipriano ([Issue 35](https://github.com/mangstadt/ez-vcard/issues/35))  
[cmargenau](https://github.com/cmargenau) ([XML 1.1 support](https://github.com/mangstadt/ez-vcard/issues/29))  
[DerBlade](https://github.com/DerBlade) ([missing parameter method](https://github.com/mangstadt/ez-vcard/issues/52))  
Ed Developer (wiki fix)  
Juliane Dombrowski ([quoted-printable line folding](https://github.com/mangstadt/ez-vcard/issues/9))  
F. Gaffron ([quoted-printable charsets](https://github.com/mangstadt/ez-vcard/issues/12))  
Edward Hetherington ([plain text KEY issue](https://github.com/mangstadt/ez-vcard/issues/91))  
[isindir](https://github.com/isindir) ([Javadoc fix](https://github.com/mangstadt/ez-vcard/pull/53))  
knutolav ([Issue 1](https://github.com/mangstadt/ez-vcard/issues/1), [Issue 2](https://github.com/mangstadt/ez-vcard/issues/2))  
[Nico Lehmann](https://github.com/ekorn) ([Windows 10 Contacts compatibility issue](https://github.com/mangstadt/ez-vcard/issues/56))  
[Alexander Myltsev](https://github.com/alexander-myltsev) ([Issue with folding quoted printable values](https://github.com/mangstadt/ez-vcard/issues/80))  
David Nault ([Issue 3](https://github.com/mangstadt/ez-vcard/issues/3), [Issue 7](https://github.com/mangstadt/ez-vcard/issues/7))  
[NeverWise](https://github.com/NeverWise) ([chaining parser & custom scribes issue](https://github.com/mangstadt/ez-vcard/issues/81))  
[rfc2822](https://github.com/rfc2822) ([folding line issue](https://github.com/mangstadt/ez-vcard/issues/30), [IMPP issue](https://github.com/mangstadt/ez-vcard/issues/32), [trailing semicolons issue](https://github.com/mangstadt/ez-vcard/issues/57), [non-Western digits](https://github.com/mangstadt/ez-vcard/issues/113))  
[Steven Ruppert](https://github.com/blendmaster) ([folding surrogate character pairs](https://github.com/mangstadt/ez-vcard/pull/36), [parsing tel URIs](https://github.com/mangstadt/ez-vcard/pull/38))  
Melchin Sapir (README suggestion)  
Matt Siegel ([base64 property value bug](https://github.com/mangstadt/ez-vcard/issues/21), [unit test bug](https://github.com/mangstadt/ez-vcard/issues/22))  
David Spieler ([hCard template bug](https://github.com/mangstadt/ez-vcard/issues/19))  
stonio ([unit test fix](https://github.com/mangstadt/ez-vcard/pull/63), [readme fix](https://github.com/mangstadt/ez-vcard/pull/68))  
Tom Vogel ([quoted-printable charsets](https://github.com/mangstadt/ez-vcard/issues/10))  
Eike Weyl (Wiki fix, Javadoc fix)  
沈健 (plain-text vCard formatting issue)

**Donators**  
[DAVDroid](https://davdroid.bitfire.at/)

**Caffeine Suppliers**  
'feine  
Starbucks  
Volo Coffeehouse

_No animals were harmed in the making of this library._

[![](https://www.paypalobjects.com/en_US/i/btn/btn_donateCC_LG.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=8CEN7MPKRBKU6&lc=US&item_name=Michael%20Angstadt&item_number=ez%2dvcard&currency_code=USD&bn=PP%2dDonationsBF%3abtn_donateCC_LG%2egif%3aNonHosted)

<a rel="me" href="https://discuss.systems/@mangstadt"></a>


history.md
Changes

Build
    bugfix HCard-Generation-geo (if locale is "de-DE")
    refactored: Replaced code-duplicates with interface default-functions
    api: added syntactical shugar: addXxx/setXxx return this to allow call-chaining
    build:
        added gradle(-submodul)-build support (parallel to pom.xml)
        updated libs (mockito:1.10.19->5.8.0; equalsverifier:1.7.6->3.15.5)
    docs: 
        updated README.md (Android compatibility, Gradle build). 
        created history.md


src/main/java/ezvcard/property/Geo.java
package ezvcard.property;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import ezvcard.VCard;
import ezvcard.VCardVersion;
import ezvcard.ValidationWarning;
import ezvcard.util.GeoUri;

/*
 Copyright (c) 2012-2023, Michael Angstadt
 All rights reserved.

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions are met: 

 1. Redistributions of source code must retain the above copyright notice, this
 list of conditions and the following disclaimer. 
 2. Redistributions in binary form must reproduce the above copyright notice,
 this list of conditions and the following disclaimer in the documentation
 and/or other materials provided with the distribution. 

 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

 The views and conclusions contained in the software and documentation are those
 of the authors and should not be interpreted as representing official policies, 
 either expressed or implied, of the FreeBSD Project.
 */

/**
 * <p>
 * Defines a set of latitude/longitude coordinates. There is no rule as to what
 * these coordinates represent, but their meaning could vary depending on the
 * value of the vCard's {@link Kind} property:
 * </p>
 * 
 * <table class="simpleTable">
 * <caption>{@link Geo} meanings by {@link Kind} value</caption>
 * <tr>
 * <th>{@link Kind} value</th>
 * <th>{@link Geo} meaning</th>
 * </tr>
 * <tr>
 * <td>"individual"</td>
 * <td>The location of the person's home or workplace.</td>
 * </tr>
 * <tr>
 * <td>"group"</td>
 * <td>The location of the group's meeting place.</td>
 * </tr>
 * <tr>
 * <td>"org"</td>
 * <td>The coordinates of the organization's headquarters.</td>
 * </tr>
 * <tr>
 * <td>"location"</td>
 * <td>The coordinates of the location itself.</td>
 * </tr>
 * </table>
 * 
 * <p>
 * <b>Code sample</b>
 * </p>
 * 
 * <pre class="brush:java">
 * VCard vcard = new VCard();
 * 
 * Geo geo = new Geo(40.7127, -74.0059);
 * vcard.setGeo(geo);
 * </pre>
 * 
 * <p>
 * <b>Property name:</b> {@code GEO}
 * </p>
 * <p>
 * <b>Supported versions:</b> {@code 2.1, 3.0, 4.0}
 * </p>
 * @author Michael Angstadt
 * @see <a href="http://tools.ietf.org/html/rfc6350#page-38">RFC 6350 p.38</a>
 * @see <a href="http://tools.ietf.org/html/rfc2426#page-16">RFC 2426 p.16</a>
 * @see <a href="http://www.imc.org/pdi/vcard-21.doc">vCard 2.1 p.16</a>
 */
public class Geo extends VCardProperty
        implements HasAltId, IPref, IV4Type, IV4Pids, IV4MediaType {
    private GeoUri uri;

    /**
     * Creates a geo property.
     *
     * @param latitude  the latitude
     * @param longitude the longitude
     */
    public Geo(Double latitude, Double longitude) {
        this(new GeoUri.Builder(latitude, longitude).build());
	}

	/**
	 * Creates a geo property.
	 * @param uri the geo URI
	 */
	public Geo(GeoUri uri) {
		this.uri = uri;
	}

	/**
	 * Copy constructor.
	 * @param original the property to make a copy of
	 */
	public Geo(Geo original) {
		super(original);
		uri = original.uri;
	}

	/**
	 * Gets the latitude.
	 * @return the latitude
	 */
	public Double getLatitude() {
		return (uri == null) ? null : uri.getCoordA();
	}

	/**
	 * Sets the latitude.
	 * @param latitude the latitude
	 */
	public void setLatitude(Double latitude) {
		if (uri == null) {
			uri = new GeoUri.Builder(latitude, null).build();
		} else {
			uri = new GeoUri.Builder(uri).coordA(latitude).build();
		}
	}

	/**
	 * Gets the longitude.
	 * @return the longitude
	 */
	public Double getLongitude() {
		return (uri == null) ? null : uri.getCoordB();
	}

	/**
	 * Sets the longitude.
	 * @param longitude the longitude
	 */
	public void setLongitude(Double longitude) {
		if (uri == null) {
			uri = new GeoUri.Builder(null, longitude).build();
		} else {
			uri = new GeoUri.Builder(uri).coordB(longitude).build();
		}
	}

	/**
	 * Gets the raw object used for storing the GEO information. This can be
	 * used to supplement the GEO value with additional information (such as
	 * altitude or level of accuracy). Geo URIs are only supported by vCard
	 * version 4.0. Only latitude and longitude values are used when marshalling
	 * to earlier vCard versions.
	 * @return the geo URI object or null if not set
	 * @see <a href="http://tools.ietf.org/html/rfc5870">RFC 5870</a>
	 */
	public GeoUri getGeoUri() {
		return uri;
	}

	/**
	 * Sets the raw object used for storing the GEO information. This can be
	 * used to supplement the GEO value with additional information (such as
	 * altitude or level of accuracy). Geo URIs are only supported by vCard
	 * version 4.0. Only latitude and longitude values are used when marshalling
	 * to earlier vCard versions.
	 * @param uri the geo URI object
	 * @see <a href="http://tools.ietf.org/html/rfc5870">RFC 5870</a>
	 */
	public void setGeoUri(GeoUri uri) {
		this.uri = uri;
	}

	@Override
	protected void _validate(List<ValidationWarning> warnings, VCardVersion version, VCard vcard) {
		if (getLatitude() == null) {
			warnings.add(new ValidationWarning(13));
		}
		if (getLongitude() == null) {
			warnings.add(new ValidationWarning(14));
		}
	}

	@Override
	protected Map<String, Object> toStringValues() {
		Map<String, Object> values = new LinkedHashMap<>();
		values.put("uri", uri);
		return values;
	}

	@Override
	public Geo copy() {
		return new Geo(this);
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = super.hashCode();
		result = prime * result + ((uri == null) ? 0 : uri.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) return true;
		if (!super.equals(obj)) return false;
		Geo other = (Geo) obj;
		if (uri == null) {
			if (other.uri != null) return false;
		} else if (!uri.equals(other.uri)) return false;
		return true;
	}
}
